sql/analyzer/stored_procedures.go

```diff
@@ -254,10 +251,105 @@ func applyProcedures(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, se
 		return n, transform.SameTree, nil
 	}
 
-	scope, err := loadStoredProcedures(ctx, a, n, scope, sel)
+	call, _, err := transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		call, ok := n.(*plan.Call)
+		if !ok {
+			return n, transform.SameTree, nil
+		}
+		if scope.IsEmpty() {
+			scope = scope.withProcedureCache(NewProcedureCache())
+		}
+		if call.AsOf != nil && !scope.enforceReadOnly {
+			scope.enforceReadOnly = true
+			defer func() {
+				scope.enforceReadOnly = false
+			}()
+		}
+
+		esp, err := a.Catalog.ExternalStoredProcedure(ctx, call.Name, len(call.Params))
+		if err != nil {
+			return nil, transform.SameTree, err
+		}
+		if esp != nil {
+			externalProcedure, err := resolveExternalStoredProcedure(ctx, *esp)
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+			return call.WithProcedure(externalProcedure), transform.NewTree, nil
+		}
+
+		if spdb, ok := call.Database().(sql.StoredProcedureDatabase); ok {
+			procedure, ok, err := spdb.GetStoredProcedure(ctx, call.Name)
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+			if !ok {
+				err := sql.ErrStoredProcedureDoesNotExist.New(call.Name)
+				if call.Database().Name() == "" {
+					return nil, transform.SameTree, fmt.Errorf("%w; this might be because no database is selected", err)
+				}
+				return nil, transform.SameTree, err
+			}
+			parsedProcedure, err := parse.Parse(ctx, procedure.CreateStatement)
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+			parsedProcedure, _, err = transform.Node(parsedProcedure, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+				switch n := n.(type) {
+				case *plan.Call:
+					newCall := n
+					tree := transform.SameTree
+					if newCall.Database() == nil || newCall.Database().Name() == "" {
+						newNode, err := newCall.WithDatabase(spdb)
+						if err != nil {
+							return nil, transform.SameTree, err
+						}
+						newCall = newNode.(*plan.Call)
+						tree = transform.NewTree
+					}
+					if newCall.AsOf == nil {
+						newCall = newCall.WithAsOf(call.AsOf)
+						tree = transform.NewTree
+					}
+					if tree == transform.NewTree {
+						return newCall, transform.NewTree, nil
+					}
+				case *plan.ShowCreateTable:
+					if n.AsOf == nil {
+						return n.WithAsOf(call.AsOf), transform.NewTree, nil
+					}
+				case *plan.ShowTables:
+					if n.AsOf == nil {
+						return n.WithAsOf(call.AsOf), transform.NewTree, nil
+					}
+				case sql.UnresolvedTable:
+					if n.AsOf() == nil {
+						newNode, err := n.WithAsOf(call.AsOf)
+						return newNode, transform.NewTree, err
+					}
```

Comment:
ooooh... very nice edge case. Totally makes sense that if a procedure is already locking the revision of a table that we'd continue to respect that and not override it. Would be nice to have a quick test to cover this case to make sure it doesn't break in the future.

Reply:
I'll add a test case!